sources/packages/api/src/services/education-program-offering/education-program-offering-validation.service.ts

```diff
@@ -0,0 +1,131 @@
+import { Injectable } from "@nestjs/common";
+import { OfferingStatus } from "../../database/entities";
+import {
+  OfferingValidationResult,
+  SaveOfferingModel,
+  ValidationWarning,
+  ValidationWarningResult,
+} from "./education-program-offering-validation.models";
+import { validateSync, ValidationError } from "class-validator";
+import { plainToClass } from "class-transformer";
+import { flattenErrors } from "../../utilities/class-validation";
+import { CustomNamedError } from "../../utilities";
+import { OFFERING_VALIDATION_CRITICAL_ERROR } from "../../constants";
+
+/**
+ * Executes the validations on an offering model.
+ */
+@Injectable()
+export class EducationProgramOfferingValidationService {
+  validateOfferingModel(
+    offering: SaveOfferingModel,
+    silently = false,
+  ): OfferingValidationResult {
+    const [validation] = this.validateOfferingModels([offering], silently);
+    return validation;
+  }
+
+  /**
+   * Validate offering models and provide the result for every model.
+   * @param offerings offering models to be validated.
+   * @param silently if true, no exception is generated case the validation fails
+   * and the success or failure can be determined from the result object.
+   * @returns validation result or an exception in the case of a failed validation
+   * combined with the silently parameter defined as false.
+   */
+  validateOfferingModels(
+    offerings: SaveOfferingModel[],
+    silently = false,
+  ): OfferingValidationResult[] {
+    return offerings.map((offering, index) => {
+      // Ensures that the object received is a class. This is needed to the
+      // proper validation metadata be available to the validation be performed.
+      const offeringModel = plainToClass(SaveOfferingModel, offering, {
+        enableImplicitConversion: true,
+      });
+      const validationsErrors = validateSync(offeringModel);
+      const allErrors: string[] = [];
+      const allWarnings: ValidationWarningResult[] = [];
+      validationsErrors.forEach((error) => {
+        const [errors, warnings] = this.extractErrorsAndWarnings(error);
+        allErrors.push(...errors);
+        allWarnings.push(...warnings);
+      });
+
+      const offeringStatus = this.getOfferingSavingStatus(
```

Comment:
Just to remind here, For offering status of the offerings which are approved by ministry, the offeringStatus will be read directly from db and we can remove this property offeringStatusToDisplay from the get dto as the form.io will not redefine this value based on the payload.

Reply:
Yes, in the future we can remove the entire calculation for the status from the form.io.